src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java

```diff
@@ -77,6 +91,37 @@ public void rebuildSessionFromFrame(MenuSession menuSession) throws CommCareSess
                         break;
                     }
                 }
+            } else if (screen instanceof QueryScreen) {
+                QueryScreen queryScreen = (QueryScreen)screen;
+                RemoteQueryDatum neededDatum = (RemoteQueryDatum) menuSession.getSessionWrapper().getNeededDatum();
+                boolean done = false;
+                for (StackFrameStep step : steps) {
+                    if (step.getId().equals(neededDatum.getDataId())) {
+                        URI uri = null;
+                        try {
+                            uri = new URI(step.getValue());
+                        } catch (URISyntaxException e) {
+                            e.printStackTrace();
+                            throw new CommCareSessionException("Query URL format error: " + e.getMessage());
+                        }
+                        try {
+                            ExternalDataInstance searchDataInstance = caseSearchHelper.getRemoteDataInstance(
```

Comment:
Is this performing the I/O synchronously inside of this call?

I think that burying the error handling here as such will cause issues long term. Previously this method's failures were all within its control, while now they aren't deterministic and could depend on, say, rate limiting from HQ.

This is kind of true in the UX version of these queries as well, but the difference is that there the I/O is in clear response to a user action, and is clearly idempotent. If it fails, it's in response to a user button press, and pressing the button again will produce the expected result.

In this case, it's not obvious to me how an I/O failure gets routed for a user, but I'm guessing it's "awkwardly." End-of-form nav isn't an obviously repeatable action, but the way the failure is routed should make it clear and possible for this to fail in an "Expected" way.

I think at the very least, the error handling should split here between structural failure (IE: there was an unrecoverable problem where XML payloads were un-parseable, repeating the action would be meaningless) and I/O failure (HQ responded with a 503) so that it's possible to differentiate both at a human level, and to let Formplayer try and do something smarter or more specific in response.